consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,459 @@
+					logger.Info("Remote validator to proxy assignment has changed.  Sending val enode share messages and updating announce version")
+					ph.sendValEnodeShareMsgs()
+					updateAnnounceVersionInFuture()
+				}
+			}
+
+		case disconnectedPeer := <-ph.removeProxyPeer:
+			// Proxied peer just disconnected.
+			peerID := disconnectedPeer.Node().ID()
+			if ph.ps.getProxy(peerID) != nil {
+				logger.Debug("Disconnected proxy peer", "peerID", peerID, "chan", "removeProxyPeer")
+				ph.ps.removeProxyPeer(peerID)
+			}
+
+		case proxyHandlerOp := <-ph.proxyHandlerOpCh:
+			proxyHandlerOp(ph.ps)
+			ph.proxyHandlerOpDoneCh <- struct{}{}
+
+		case <-ph.newBlockchainEpoch:
+			// New blockchain epoch. Update the validators in the proxySet
+			valsReassigned, error := ph.updateValidators()
+			if error != nil {
+				logger.Warn("Error in updating validator assignments on new epoch", "error", error)
+			}
+			if valsReassigned {
+				ph.sendValEnodeShareMsgs()
+				updateAnnounceVersionInFuture()
+			}
+
+		case <-updateAnnounceVersionTimer.C:
+			// updateAnnounceVersionRequests should never be empty here.
+			// If it is, there is a bug in the code
+			if len(updateAnnounceVersionRequestTimestamps) == 0 {
+				logger.Error("updateAnnounceVersionRequestTimestamps is empty when updateAnnounceVersionTimer expired")
+			} else {
+				now := time.Now()
+				updateSent := false
+				numRequestToPop := 0
+
+				for _, minRequestTimestamp := range updateAnnounceVersionRequestTimestamps {
+					if minRequestTimestamp.Before(now) || minRequestTimestamp.Equal(now) {
+						if !updateSent {
+							ph.sb.UpdateAnnounceVersion()
+							updateSent = true
+						}
+
+						numRequestToPop++
+					} else {
+						// Update the timer to tick for the first entry of the requests
+						updateAnnounceVersionTimer.Reset(updateAnnounceVersionRequestTimestamps[0].Sub(now))
+						break
+					}
+				}
+
+				if numRequestToPop > 0 {
+					updateAnnounceVersionRequestTimestamps = updateAnnounceVersionRequestTimestamps[numRequestToPop:]
+				}
+			}
+
+		case <-phEpochTicker.C:
+			logger.Trace("phEpochTicker expired.  Will process disconnected proxies.")
+			// Remove validator assignement for proxies that are disconnected for at minmum `proxyHandlerEpochLength` seconds.
+			// The reason for not immediately removing the validator asssignments is so that if there is a
+			// network disconnect then a quick reconnect, the validator assignments wouldn't be changed.
+			if valsReassigned := ph.ps.unassignDisconnectedProxies(ph.proxyHandlerEpochLength); valsReassigned {
+				ph.sendValEnodeShareMsgs()
+				updateAnnounceVersionInFuture()
+			} else {
+				// This is the case if there were no changes the validator assignments at the end of `proxyHandlerEpochLength` seconds.
+
+				// Send out the val enode share message.  We will resend the valenodeshare message here in case it was
+				// never successfully sent before.
+				ph.sendValEnodeShareMsgs()
```

Comment:
This shouldn't be as big of a concern.  Previously, it was sending out a message that was O(n^2), where n is the val set size.  Now the bytes sent out will be O(n).  It will only send a subset of the val enode table to each proxy (the entries that a specific proxy is responsible for).